### PR TITLE
Adds a Colab demonstrating integration with RayTune

### PR DESCRIPTION
**Pull request by charlesfrye**
@AyushExel I'm gonna make my review for this Colab on ReviewNB.

### CODE REVIEW

File: colabs/raytune/W&B + RayTune.ipynb [hunk 1]
```diff
@@ -0,0 +1,147 @@
+{
```
Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:1]: <p>You should always start with the Weights &amp; Biases logo image:</p><p> </p><p>&lt;img&nbsp;src="https://i.imgur.com/gb6B4ig.png"&nbsp;width="400"&nbsp;alt="Weights&nbsp;&amp;&nbsp;Biases"&nbsp;/&gt;</p><p> </p><p>Followed by an H1 Header (<code>#</code> in a Markdown cell) with a title. Here, W&amp;B ♥ RayTune, or similar.</p><p> </p><p>I would then write a short paragraph about what RayTune is, what we are, and why we work so well together. Maybe crib some of the text from your report: https://www.wandb.com/articles/distributed-hyperparameter-optimization-at-scale</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/37/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:1]: <p>Line 1: It's preferred to import in three blocks: standard library modules, then 3rd-party libraries, then your stuff (if any). So move <code>random</code>  to the top with <code>getpass</code>.</p><p>Lines 5-6: It's preferred to import in alphabetical order within each block.</p><p>Line 9: This is a repeat of Line 7.</p><p> </p><p>Lines 14-16: FYI, there's a bit more to do here to make Torch reproducible:</p><p> </p><pre class="ql-syntax" spellcheck="false">#&nbsp;Ensure&nbsp;deterministic&nbsp;behavior
torch.backends.cudnn.deterministic&nbsp;=&nbsp;True
random.seed(hash("setting&nbsp;random&nbsp;seeds")&nbsp;%&nbsp;2**32&nbsp;-&nbsp;1)
np.random.seed(hash("improves&nbsp;reproducibility")&nbsp;%&nbsp;2**32&nbsp;-&nbsp;1)
torch.manual_seed(hash("by&nbsp;removing&nbsp;stochasticity")&nbsp;%&nbsp;2**32&nbsp;-&nbsp;1)
torch.cuda.manual_seed_all(hash("so&nbsp;runs&nbsp;are&nbsp;repeatable")&nbsp;%&nbsp;2**32&nbsp;-&nbsp;1)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/37/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:1]: <p>Reading this, it's a bit unclear to me what's going on here. Do I always need to include both the <code>Logger</code>  and the <code>mixin</code>?  Is one more flexible than the other? Give a bit more detail here.</p><p> </p><p>Spread the code out, if you can, and interleave with text explaining what's going on, why you're doing it, and what else you could've done, if the situation were different.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/37/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:1]: <p>Line 4: Love the inclusion of a docstring. The note about <code>wandb.log</code> being allowed only within its scope is great, but I think that should really be in the body of the text outside the cells, as you describe what the mixin is for.</p><p> </p><p>Line 7-8: I might move the imports of the example utilities down from the top of the notebook and closer to where they get used. Specifically note that you're using them, maybe mention that they're useful for debugging, link to any documentation.</p><p> </p><p>Line 10: put a space after the comma between <code>["lr"]</code> and <code>momentum</code></p><p> </p><p>Line 12: I like that you've fixed the number of iterations to be 10 to keep runtime down!</p><p> </p><p>Line 13: put a space after the comma between <code>train_loader</code> and <code>device</code></p><p> </p><p>Line 14: put a space after the comma between <code>test_loader</code> and <code>device</code></p><p> </p><p>Line 16: I suggest using "Error Rate" as the custom metric: <code>100 * (1 - acc)</code>. It's got a meaning, but it's still trivial to calculate. You'd update the name in Line 18.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/37/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:1]: <p>I would split out the part where you set the config from the part where you give it to <code>tune</code>.  That's a chance to talk about what goes in a config, link to their docs, link to our <code>wandb.init</code>  docs.</p><p> </p><p>Lines 4 &amp; 6: I think we should drop these newlines.</p><p>Line 11: Why is <code>monitor_gym</code>  here? There's no gym.</p><p>Line 11: <code>raytune</code>  rather than <code>rayTune</code> .</p><p>Lines 14-15: These are nice targets for a <code>grid_search</code>,  but the search needs some adjustment. First, you duplicate the <code>lr</code> argument <code>0.1</code>  twice. <code>0.1</code>  is the one true learning rate, but it shouldn't be repeated. Also, they're not in order and there's no obvious rule to their spacing. This applies to <code>momentum</code>  as well. Finally, the search is too long: 24 runs is too many just for a demo.</p><p> </p><p>I suggest you switch to:</p><p> </p><pre class="ql-syntax" spellcheck="false">&nbsp;&nbsp;&nbsp;&nbsp;"lr": tune.grid_search([0.0001, 0.001, 0.1]),
&nbsp;&nbsp;&nbsp;&nbsp;"momentum": tune.grid_search([0.9, 0.99])
</pre><p> </p><p>Don't forget those spaces after the commas!&nbsp;&nbsp;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/37/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

**Comment by charlesfrye** [comment]
@AyushExel nice work! The example is great, just recommending some minor adjustments. I'd love to see more explanatory text as well.

Managing the workflow with your permissions is tricky. I suggest this:
1) Fork the examples repo to your own account.
2) Move to the `colabs/raytune-torch` branch (hereafter: the "feature branch") on the fork, then head to the NB file in that fork.
2) Click the "Open in Colab" link on the top of the file
3) Make the requested edits, saving things to your Drive in the interim.
4) When you're done, save it onto the feature branch on your fork.
5) Make a PR onto the original repo's feature branch to we can see the diff in ReviewNB.
6) Once that's good to go, I'll merge your fork's feature branch into the original repo's feature branch, then into master.

In the future, I think I'll ask that folks adding Colabs make PRs from repos they control (eg forks) to make this process more streamlined.

File: colabs/raytune/W&B + RayTune.ipynb (hunk 1 above)
Comment (AyushExel) [reply colabs/raytune/W&B + RayTune.ipynb:1]: I didn't know about `torch.backends.cudnn.deterministic`. Thanks

**Comment by charlesfrye** [comment]
superceded by #40
